### ISSUE TITLE
Define the interface of Dataset.py

### ISSUE THREAD

**Issue by MarcCote** [SMART-Lab/smartlearner#4]
We should discuss here the interface of the datasets the smartpy library will be manipulating. In contrast with https://github.com/SMART-Lab/mldata, this is not a generic dataset manager.

**Some questions**
- [ ] What type of dataset we want to support?
- [ ] How will we support the notion of inputs and targets (i.e. variable number of sets of data)?
- [ ] Are trainset, validset and testset only three instances of `Dataset` or the split should be part of the class?
- [x] Should we provide symbolic variable for inputs and targets? Yes, it is a convenient place to put them their as they would be easy accessible instead of behind hidden in some obscure function responsible to compile the Theano graph.
- [ ] How do we support targets for unlabeled dataset? Usually, models using unlabeled dataset use the input as their target, should this be the default behaviour when there are no targets?

**Comment by mgermain** [comment]
I think you are pretty much asking the question we were not exactly capable to answer in mldata. I think for starters we should only support the basic kind for classification with a tuple (input, target) where input is an array and target a onehot vector. After that, we could do a small modification and support the "unsupervised kind" (input, input) and slowly but surely add support for new dataset types.

I suggest that we only focus on answering "Are trainset, validset and testset only three instances of Dataset or the split should be part of the class?", that we will need for the basic case, for now.

**Comment by MarcCote** [comment]
I agree we should not try to be too general nor too specific. I say target should simply be another array (not necessarily a one-hot vector!).

Regarding the question "Are trainset, validset and testset only three instances of Dataset or the split should be part of the class?", the way I see it is: trainset, validset and testset are sets of data, so they should be different instances of `Dataset`. But I don't know how to named the MNIST, CalTech101, CIFAR, etc. "datasets"? Maybe `DatasetCollection`?

**Comment by mgermain** [comment]
Good point for the one-hot. As for the valid train test I see them more as
views over a given dataset.
